core/math/a_star_grid_2d.h

```diff
@@ -0,0 +1,207 @@
+		DIAGONAL_MODE_ONLY_IF_NO_OBSTACLES,
+		DIAGONAL_MODE_MAX,
+	};
+
+	enum Heuristic {
+		HEURISTIC_EUCLIDEAN,
+		HEURISTIC_MANHATTAN,
+		HEURISTIC_OCTILE,
+		HEURISTIC_CHEBYSHEV,
+		HEURISTIC_MAX,
+	};
+
+private:
+	int64_t width = 0;
+	int64_t height = 0;
+
+	bool jumping_enabled = false;
+	DiagonalMode diagonal_mode = DIAGONAL_MODE_ALWAYS;
+	Heuristic default_heuristic = HEURISTIC_EUCLIDEAN;
+
+	enum Dir {
+		DIR_LEFT,
+		DIR_RIGHT,
+		DIR_TOP,
+		DIR_TOP_LEFT,
+		DIR_TOP_RIGHT,
+		DIR_BOTTOM,
+		DIR_BOTTOM_LEFT,
+		DIR_BOTTOM_RIGHT,
+	};
+
+	struct Point {
+		int64_t id = 0;
+
+		_FORCE_INLINE_ int64_t get_x(int64_t p_grid_width) { return id % p_grid_width; }
+		_FORCE_INLINE_ int64_t get_y(int64_t p_grid_width) { return id / p_grid_width; }
+
+		real_t weight_scale = 0;
+		bool enabled = true;
+		Vector2 pos;
+
+		Point *neighbours[8] = { nullptr };
+
+		// Used for pathfinding.
+		Point *prev_point = nullptr;
+		real_t g_score = 0;
+		real_t f_score = 0;
+		uint64_t open_pass = 0;
+		uint64_t closed_pass = 0;
+
+		Point() {}
+	};
+
+	struct JumpState {
+		int64_t from;
+		int64_t to;
+		int stage = 0;
+	};
+
+	struct SortPoints {
+		_FORCE_INLINE_ bool operator()(const Point *A, const Point *B) const { // Returns true when the Point A is worse than Point B.
+			if (A->f_score > B->f_score) {
+				return true;
+			} else if (A->f_score < B->f_score) {
+				return false;
+			} else {
+				return A->g_score < B->g_score; // If the f_costs are the same then prioritize the points that are further away from the start.
+			}
+		}
+	};
+
+	OAHashMap<int64_t, Point *> points = 0U;
+	Point *end = nullptr;
+
+	int64_t last_free_id = 0;
+	uint64_t pass = 1;
+
+private: // Internal routines.
+	_FORCE_INLINE_ int64_t _idx(int64_t p_x, int64_t p_y) const {
+		return p_y * width + p_x;
+	}
+
+	_FORCE_INLINE_ bool _is_walkable(int64_t p_x, int64_t p_y) const {
+		if (p_x >= 0 && p_y >= 0 && p_x < width && p_y < height) {
+			Point *pt = (*points.lookup_ptr(_idx(p_x, p_y)));
+			return pt->enabled;
+		}
+		return false;
+	}
+
+	_FORCE_INLINE_ bool _is_walkable_nbor(Point *p_point, int p_key, Point **r_point) const {
+		Point *pt = p_point->neighbours[p_key];
+		if (pt != nullptr) {
+			(*r_point) = pt;
+			return pt->enabled;
+		}
+		return false;
+	}
+
+	_FORCE_INLINE_ Point *_get_point(int64_t p_x, int64_t p_y) {
+		Point *pt;
+		if (points.lookup(_idx(p_x, p_y), pt)) {
+			return pt;
+		}
+		return nullptr;
+	}
+
+	void _get_nbors(Point *p_point, List<Point *> &r_nbors);
+	Point *_jump(Point *p_from, Point *p_to);
+	bool _solve(Point *p_begin_point, Point *p_end_point);
+
+protected:
+	static void _bind_methods();
+
+	virtual real_t _estimate_cost(int64_t p_from_id, int64_t p_to_id);
+	virtual real_t _compute_cost(int64_t p_from_id, int64_t p_to_id);
+
+	GDVIRTUAL2RC(real_t, _estimate_cost, int64_t, int64_t)
+	GDVIRTUAL2RC(real_t, _compute_cost, int64_t, int64_t)
+
+public:
+	void create_grid(int64_t p_width, int64_t p_height, real_t p_weight_scale);
```

Comment:
I think this should simply be a `set_size(Size2i size) and Size2i get_size()` and have an exposed `size` property.